File: dataPreprocessing_v4/ADASSegment.py

class ADASSegment:
    def __init__(self, i, dca, rad, lan, st, len, limv, time, gps):
        self.idx = i
        self.raw_dir_ch_angle = dca
        self.raw_rad = rad
        self.lanes = lan
        self.street_type = st
        self.length = len
        self.lim_v = limv
        self.time = time
        self.gps_coord = gps

    def __str__(self):
        return self.print_data()

    def print_data(self):
        return f"idx[{self.print_fena_ffinv(self.idx)}] " \
               f"dca[{self.print_dca(self.raw_dir_ch_angle)}] " \
               f"rad[{self.print_rad(self.raw_rad)}] " \
               f"lan[{self.print_lanes(self.lanes)}] " \
               f"st[{self.street_type}] " \
               f"len[{self.print_fena_ffinv(self.length)}] " \
               f"limv[{self.print_lim_v(self.lim_v)}] " \
               f"time[{self.time}]"

    @staticmethod
    def print_fena_ffinv(val):
        if val == 254:
            return "n.a."
        if val == 255:
            return "inv."
        return str(val)

    def print_dca(self, dca):
        if dca == 4094:
            self.dir_ch_angle = None
            return "n.a."
        if dca == 4095:
            self.dir_ch_angle = None
            return "inv."
        self.dir_ch_angle = float(dca) * 0.08793356 - 180.0
        return f"{self.dir_ch_angle:.1f}"

    def print_rad(self, rad):
        if rad == 4093:
            self.rad = None
            return "straight"
        if rad == 4094:
            self.rad = None
            return "n.a."
        if rad == 4095:
            self.dir_ch_angle = None
            return "inv."
        self.rad = rad * 5 - 10230
        return str(self.rad)

    @staticmethod
    def print_lanes(lan):
        if lan == 0:
            return ">5"
        if lan == 6:
            return "n.a."
        if lan == 7:
            return "inv."
        return str(lan)

    @staticmethod
    def print_lim_v(limv):
        if limv == 0:
            return "unlim"
        if limv == 150:
            return "n.a."
        if limv == 155:
            return "inv."
        return str(limv)

File: 2023.04.28_asc2kml/src/asc2kml/ADASSegment.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package asc2kml;

import static java.lang.Double.NaN;

/**
 *
 * @author Q262579
 */
public class ADASSegment {

    public GPSCoordinate gpsCoord;
    public long reltime;
    public int Idx;
    public int RawDirChAngle;
    public double DirChAngle;
    public int RawRad;
    public int Rad;
    public int Lanes;
    public STREETTYPE StreetType;
    public int Length;
    public int MapProps;
    public boolean urban;
    public boolean crossing;
    public boolean roundabout;
    public boolean parking;
    public boolean ramp;
    public boolean trafficlight;
    public boolean urbanlimit;
    public boolean signlimit;
    public boolean righthanddrive;
    public boolean generallimit;
    public int LimV;


    public ADASSegment(int i, int dca, int rad, int lan, STREETTYPE st, int len, int propm, int limv, GPSCoordinate gps, long rt) {
        Idx = i;
        RawDirChAngle = dca;
        RawRad = rad;
        Lanes = lan;
        StreetType = st;
        Length = len;
        MapProps = propm;
        setMapProps(MapProps);        
        LimV = limv;
        gpsCoord = gps;
        reltime = rt;
    }
    public ADASSegment(ADASSegment as) {
        this.Idx = as.Idx;
        this.DirChAngle = as.DirChAngle;
        this.RawDirChAngle = as.RawDirChAngle;
        this.Rad = as.Rad;
        this.RawRad = as.RawRad;
        this.Lanes = as.Lanes;
        this.StreetType = as.StreetType;
        this.Length = as.Length;
        this.MapProps = as.MapProps;
        this.setMapProps(as.MapProps);
        this.LimV = as.LimV;
        this.gpsCoord = new GPSCoordinate(as.gpsCoord);
        this.reltime = as.reltime;
    }
    @Override
    public String toString() {
        return listAllData();
    }
    public String listIndex() {
        return "t[" + reltime + "] " +
               "idx[" + printFEna_FFinv(Idx) + "]";
    }
    public String listAllData() {
        return "t[" + reltime + "] " +
               "idx[" + printFEna_FFinv(Idx) + "] " +
               "dca[" + printDCA(RawDirChAngle) + "] " +
               "rad[" + printRAD(RawRad) + "] " +
               "lan[" + printLanes(Lanes) + "] " +
               "st[" + StreetType.toString() + "] " +
               "len[" + printFEna_FFinv(Length) + "] " +
               "map[" +  printMapProps() + "] " +
               "limv[" + printLimV(LimV) + "] ";
    }
    private boolean getBit(int val, int mask) {        
        return (val & mask) == mask;
    }
    private void setMapProps(int mp) {        
        urban = getBit(mp, 1);
        crossing = getBit(mp, 2);
        roundabout = getBit(mp, 4);
        parking = getBit(mp, 8);
        ramp = getBit(mp, 16);
        trafficlight = getBit(mp, 32);
        urbanlimit = getBit(mp, 64);
        signlimit = getBit(mp, 128);
        righthanddrive = getBit(mp, 512);
        generallimit = getBit(mp, 1024);        
    }
    private String printMapProps() {
        String mp = "";
        
        if (MapProps == 2046) // not available
            return "//////////";        
        if (MapProps == 2047) // invalid
            return "!!!!!!!!!!";
        
        mp += urban ? "U" : "-";
        mp += crossing ? "X" : "-";
        mp += roundabout ? "O" : "-";
        mp += parking ? "P" : "-";
        mp += ramp ? "R" : "-";
        mp += trafficlight ? "L" : "-";
        mp += urbanlimit ? "u" : "-";
        mp += signlimit ? "s" : "-";
        mp += righthanddrive ? "r" : "-";
        mp += generallimit ? "g" : "-";
        
        return mp;
    }
    private String printFEna_FFinv(int val) {        
        if (val == 254)
            return "na";
        if (val == 255)
            return "inv";
        return Integer.toString(val);
    }
    private String printDCA(int dca) {
        if (dca == 4094) {
            DirChAngle = NaN;
            return "na";
        }
        if (dca == 4095) {
            DirChAngle = NaN;
            return "inv";
        }
        DirChAngle = (double)dca * 0.08793356 - 180.0;
        return String.format("%.1f",DirChAngle);
    }
    private String printRAD(int rad) {
        if (rad == 4093) {
            Rad = (int) NaN;
            return "straight";
        }
        if (rad == 4094) {
            Rad = (int) NaN;
            return "na";
        }
        if (rad == 4095) {
            DirChAngle = (int) NaN;
            return "inv";
        }
        Rad = rad * 5 - 10230;
        return Integer.toString(Rad);
    }
    private String printLanes(int lan) {
        if (lan == 0)
            return ">5";
        if (lan == 6)
            return "na";
        if (lan == 7)
            return "inv";
        return Integer.toString(lan);
    }
    private String printLimV(int limv) {
        if (limv == 0)
            return "unlim";
        if (limv == 150)
            return "na";
        if (limv == 155)
            return "inv";
        return Integer.toString(limv);
    }
}
